tests/test_worker.py

```diff
@@ -0,0 +1,857 @@
+        async with temporalio.worker.Worker(
+            client, task_queue=act_task_queue, activities={"say_hello": say_hello}
+        ):
+            await client.execute_workflow(
+                "kitchen_sink",
+                KSWorkflowParams(
+                    actions=[
+                        KSAction(
+                            execute_activity=KSExecuteActivityAction(
+                                name="wrong_activity", task_queue=act_task_queue
+                            )
+                        )
+                    ]
+                ),
+                id=str(uuid.uuid4()),
+                task_queue=worker.task_queue,
+            )
+    assert str(assert_activity_application_error(err.value)) == (
+        "Activity function wrong_activity is not registered on this worker, "
+        "available activities: say_hello"
+    )
+
+
+async def test_max_concurrent_activities(
+    client: temporalio.client.Client, worker: Worker
+):
+    seen_indexes: List[int] = []
+    complete_activities_event = asyncio.Event()
+
+    async def some_activity(index: int) -> str:
+        seen_indexes.append(index)
+        # Wait here to hold up the activity
+        await complete_activities_event.wait()
+        return ""
+
+    # Only allow 42 activities, but try to execute 43. Make a short schedule to
+    # start timeout but a long schedule to close timeout.
+    with pytest.raises(temporalio.client.WorkflowFailureError) as err:
+        await _execute_workflow_with_activity(
+            client,
+            worker,
+            some_activity,
+            count=43,
+            index_as_arg=True,
+            schedule_to_close_timeout_ms=5000,
+            schedule_to_start_timeout_ms=1000,
+            worker_config={"max_outstanding_activities": 42},
+            on_complete=complete_activities_event.set,
+        )
+    timeout = assert_activity_error(err.value)
+    assert isinstance(timeout, temporalio.exceptions.TimeoutError)
+    assert str(timeout) == "activity timeout"
+    assert timeout.type == temporalio.exceptions.TimeoutType.SCHEDULE_TO_START
+
+
+@dataclass
+class SomeClass1:
+    foo: int
+
+
+@dataclass
+class SomeClass2:
+    foo: str
+    bar: Optional[SomeClass1] = None
+
+
+async def test_activity_type_hints(client: temporalio.client.Client, worker: Worker):
+    activity_param1: SomeClass2
+
+    async def some_activity(param1: SomeClass2, param2: str) -> str:
+        nonlocal activity_param1
+        activity_param1 = param1
+        return f"param1: {type(param1)}, param2: {type(param2)}"
+
+    result = await _execute_workflow_with_activity(
+        client,
+        worker,
+        some_activity,
+        SomeClass2(foo="str1", bar=SomeClass1(foo=123)),
+        123,
+    )
+    # We called with the wrong non-dataclass type, but since we don't strictly
+    # check non-data-types, we don't perform any validation there
+    # TODO(cretz): Do we want a strict option for scalars?
```

Comment:
Can have discussion on how we use type hinting. Right now, to be the most Pythonic and flexible, we _only_ use it when converting from JSON payloads to dataclasses. Otherwise we ignore it and therefore allow it to be wrong just like Python. But we could be stricter if there was value.

Reply:
Meh, this discussion can be had at a later time.